rust/src/tree.rs

```diff
@@ -0,0 +1,215 @@
+use std::collections::HashMap;
+use std::mem;
+
+#[derive(Clone)]
+pub struct Node {
+    pub children: HashMap<usize, Node>, // the key is first id of the child because each child must have unique first id
+    pub ids: Vec<usize>,
+    pub count: usize,
+}
+
+pub struct RadixTree {
+    pub root: Node,
+}
+
+fn common_prefix_len(a: &[usize], b: &[usize]) -> usize {
+    let mut i = 0;
+    while i < a.len() && i < b.len() && a[i] == b[i] {
+        i += 1;
+    }
+    i
+}
+
+impl Default for RadixTree {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl RadixTree {
+    pub fn new() -> Self {
+        RadixTree {
+            root: Node {
+                children: HashMap::new(),
+                ids: Vec::new(),
+                count: 0,
+            },
+        }
+    }
+
+    pub fn insert(&mut self, input_ids: &[usize]) {
+        let mut curr = &mut self.root;
+        curr.count += 1;
+
+        let mut curr_idx = 0;
+        let input_ids_len = input_ids.len();
+
+        while curr_idx < input_ids_len {
+            let mut match_info = None;
+
+            for key in curr.children.keys() {
```

Comment:
Since you have changed the children to be a hashmap. The for loop is not needed here.